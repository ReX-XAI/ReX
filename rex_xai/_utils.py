#!/usr/bin/env python3

import importlib.metadata
from itertools import chain, combinations
from enum import Enum
from typing import Tuple, Union, Dict
from numpy.typing import NDArray
import torch as tt
import numpy as np
from skimage.segmentation import mark_boundaries
from rex_xai.logger import logger
from rex_xai.box import Box

Strategy = Enum("Strategy", ["Global", "Spatial", "MultiSpotlight", "Contrastive"])

Queue = Enum("Queue", ["Area", "All", "Intersection", "DC"])

SpatialSearch = Enum("SpatialSearch", ["NotFound", "Found"])


def powerset(r, reverse=True):
    ps = list(chain.from_iterable(combinations(r, lim) for lim in range(1, len(r) + 1)))
    if reverse:
        return reversed(ps)
    else:
        return ps


def clause_area(clause, areas: Dict) -> int:
    tot = 0
    for c in clause:
        tot += areas[c]
    return tot


class ReXError(Exception):
    pass


class ReXTomlError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXTomlError: {self.message}"


class ReXPathError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXPathError: no such file exists at {self.message}"


class ReXScriptError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXScriptError: {self.message}"


class ReXDataError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXDataError: {self.message}"


class ReXMapError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXMapError: {self.message}"


def xlogx(ps):
    f = np.vectorize(_xlogx)
    return f(ps)


def _xlogx(p):
    if p == 0.0:
        return 0.0
    else:
        return p * np.log2(p)


def add_boundaries(
    img: Union[NDArray, tt.Tensor], segs: NDArray, colour=None
) -> NDArray:
    if colour is None:
        m = mark_boundaries(img, segs, mode="thick")
    else:
        m = mark_boundaries(img, segs, colour, mode="thick")
    m *= 255  # type: ignore
    m = m.astype(np.uint8)
    return m


def get_device(gpu: bool):
    if tt.backends.mps.is_available() and gpu:
        return tt.device("mps")
    if tt.device("cuda") and gpu:
        return tt.device("cuda")
    if gpu:
        logger.warning("gpu not available")
    return tt.device("cpu")


def get_map_locations(map, reverse=True):
    if isinstance(map, tt.Tensor):
        map = map.detach().cpu().numpy()
    coords = []
    for i, r in enumerate(np.nditer(map)):
        coords.append((r, np.unravel_index(i, map.shape)))
    coords = sorted(coords, reverse=reverse)
    return coords


def set_boolean_mask_value(
    tensor,
    mode,
    order,
    coords: Union[Box, Tuple[NDArray, NDArray]],
    val: bool = True,
):
    if isinstance(coords, Box):
        if mode in ("spectral", "tabular"):
            h = coords.col_start
            w = coords.col_stop
        else:
            h = slice(coords.row_start, coords.row_stop)
            w = slice(coords.col_start, coords.col_stop)
    else:
        h = coords[0]
        w = coords[1]
    # three channels
    if mode == "RGB":
        # (C, H, W)
        if order == "first":
            tensor[:, h, w] = val
        # (H, W, C)
        else:
            tensor[h, w, :] = val
    elif mode == "L":
        if order == "first":
            # (1, H, W)
            tensor[0, h, w] = val
        else:
            tensor[h, w, :] = val
    elif mode in ("spectral", "tabular"):
        if len(tensor.shape) == 1:
            tensor[h:w] = val
        else:
            tensor[0, h:w] = val
    # elif mode == "tabular":

    elif mode == "voxel":
        logger.warning("not yet implemented")


def ff(obj, fmt):
    """
    Like format(obj, fmt), but returns the string 'None' if obj is None.
    See the help for format() to see acceptable values for fmt.
    """
    return "None" if obj is None else format(obj, fmt)


def version():
    return importlib.metadata.version("rex-xai")
