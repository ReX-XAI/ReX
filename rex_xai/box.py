#!/usr/bin/env python

"""generate boxes which together create a occlusion.
This occlusion is realised in the form of a mask over an image"""

from __future__ import annotations

from typing import List, Tuple, Dict
from anytree import LevelOrderGroupIter, NodeMixin, RenderTree
import numpy as np

from rex_xai.distributions import Distribution, random_coords
from rex_xai.logger import logger

# Enums for different axes
class Axes:
    ROW = 0
    COL = 1
    DEPTH = 2


class BoxInternal:
    """a box is part of an occulsion, a collection of boxes which form a mask over an image"""

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        row_start,
        row_stop,
        col_start,
        col_stop,
        depth_start=None,  # Optional depth_start and depth_stop for 3D data
        depth_stop=None,
        distribution=None,
        distribution_args=None,
        name="",
    ) -> None:
        self.name = name
        self.distribution = distribution
        self.distribution_args = distribution_args
        self.row_start: int = row_start
        self.row_stop: int = row_stop
        self.col_start: int = col_start
        self.col_stop: int = col_stop
        self.depth_start = depth_start
        self.depth_stop = depth_stop

    def __repr__(self) -> str:
        if self.depth_start is not None and self.depth_stop is not None:
            return (
                f"Box < name: {self.name}, row_start: {self.row_start}, "
                + f"row_stop: {self.row_stop}, col_start: {self.col_start}, "
                + f"col_stop: {self.col_stop}, depth_start: {self.depth_start}, "
                + f"depth_stop: {self.depth_stop}, volume: {self.area()}"
            )
        else:
            return (
                f"Box < name: {self.name}, row_start: {self.row_start}, "
                + f"row_stop: {self.row_stop}, col_start: {self.col_start}, "
                + f"col_stop: {self.col_stop}, area: {self.area()}"
            )

    def __hash__(self) -> int:
        return hash(self.name)

    def __eq__(self, other) -> bool:
        if isinstance(other, int):
            return self.name == other
        if isinstance(other, type(self)):
            return NotImplemented
        return self.name == other.name

    def update_name(self, name: str):
        """sets the name of a box"""
        self.name += name

    def shape(self):
        """returns (width, height) of a box if 2d data, else returns (width, height, depth)"""
        if self.depth_start is not None and self.depth_stop is not None:
            return (
                self.row_stop - self.row_start,
                self.col_stop - self.col_start,
                self.depth_stop - self.depth_start,
            )
        else:
            return (self.row_stop - self.row_start, self.col_stop - self.col_start)

    def corners(self):
        """Return (Wstart, Wstop, Hstart, Hstop) of current box if 2d data, else returns (Wstart, Wstop, Hstart, Hstop. Dstart, Dstop)"""
        if self.depth_start is not None and self.depth_stop is not None:
            return (
                self.row_start,
                self.row_stop,
                self.col_start,
                self.col_stop,
                self.depth_start,
                self.depth_stop,
            )
        else:
            return (self.row_start, self.row_stop, self.col_start, self.col_stop)

    def __1d_parts(self):
        c1 = random_coords(self.distribution, [self.col_stop - self.col_start])
        if c1 is not None and isinstance(c1, np.ndarray):
            c1 = c1[0] + self.col_start

        c2 = random_coords(self.distribution, [self.col_stop - self.col_start])
        if c2 is not None and isinstance(c2, np.ndarray):
            c2 = c2[0] + self.col_start

        c3 = random_coords(self.distribution, [self.col_stop - self.col_start])
        if c3 is not None and isinstance(c3, np.ndarray):
            c3 = c3[0] + self.col_start

        if c1 is None or c2 is None or c3 is None:
            return None

        ordered = sorted([c1, c2, c3])

        b0 = Box(
            0,
            0,
            self.col_start,
            ordered[0],
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b0.update_name(":0")

        b1 = Box(
            0,
            0,
            ordered[0],
            ordered[1],
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b1.update_name(":1")

        b2 = Box(
            0,
            0,
            ordered[1],
            ordered[2],
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b2.update_name(":2")

        b3 = Box(
            0,
            0,
            ordered[2],
            self.col_stop,
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b3.update_name(":3")

        return [b0, b1, b2, b3]

    def __2d_parts(self, map=None):
        if self.distribution == Distribution.Adaptive:
            pos = random_coords(self.distribution, self.corners(), map=map)
        else:
            h = int(self.row_stop - self.row_start)
            w = int(self.col_stop - self.col_start)
            space: int = h * w
            pos = random_coords(
                self.distribution, space, h, w, self.distribution_args, map=map
            )

        if pos is None:
            return
        coords = np.unravel_index(
            pos,
            (self.row_stop - self.row_start, self.col_stop - self.col_start),
        )  # type: ignore
        row: int = int(coords[0]) + self.row_start
        col: int = int(coords[1]) + self.col_start

        b0 = Box(
            self.row_start,
            row,
            self.col_start,
            col,
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b0.update_name(":0")

        b1 = Box(
            self.row_start,
            row,
            col,
            self.col_stop,
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b1.update_name(":1")

        b2 = Box(
            row,
            self.row_stop,
            self.col_start,
            col,
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b2.update_name(":2")

        b3 = Box(
            row,
            self.row_stop,
            col,
            self.col_stop,
            distribution=self.distribution,
            distribution_args=self.distribution_args,
            name=self.name,
        )
        b3.update_name(":3")

        return [b0, b1, b2, b3]

    def __3d_parts(self, map=None):
        """
        Create 4 boxes from the original box passed in as an argument.
        Pick two axes to split on randomly using the Axes class and create boxes.
        """

        # Pick two axes to split on randomly using the axes class
        axes = [Axes.ROW, Axes.COL, Axes.DEPTH]
        selected_axes = np.random.choice(axes, 2, replace=False)
        # Range of the different axes to split on
        ranges: Dict = {
            Axes.ROW: [self.row_start, self.row_stop],
            Axes.COL: [self.col_start, self.col_stop],
            Axes.DEPTH: [self.depth_start, self.depth_stop],
        }
        range1 = ranges[selected_axes[0]]
        range2 = ranges[selected_axes[1]]
        logger.debug(f"Selected axes: {selected_axes}, which have a range of values, {range1} and {range2}")
        # Get the random coordinates for the two axes
        space = range1[1] - range1[0]
        if space == 0 or space == 1:
            c1 = np.random.choice([range1[0], range1[1]])
        else:
            c1 = random_coords(
                self.distribution,
                space,
                range1[0],
                range1[1],
                self.distribution_args,
                map=map,
            )
            c1 = range1[0] + c1
        space = range2[1] - range2[0]
        if space == 0 or space == 1:
            c2 = np.random.choice([range2[0], range2[1]])
        else:
            c2 = random_coords(
                self.distribution,
                space,
                range2[0],
                range2[1],
                self.distribution_args,
                map=map,
            )
            c2 = range2[0] + c2
        logger.debug(f"Random coordinates picked: {c1} and {c2}")
        # If any of the coordinates are None, return None
        if c1 == -1 or c2 == -1:
            return None
        # Create boxes depending on the selected axes
        boxes = []

        subboxes = self.create_box(selected_axes[0], c1, self)
        for i, box in enumerate(subboxes):
            boxes.extend(self.create_box(selected_axes[1], c2, box))
        # Update the names of the boxes
        for i, box in enumerate(boxes):
            box.update_name(f":{i}")
        return boxes

    def create_box(self, axes, c1, box: BoxInternal):
        """
        Create a box depending on the selected axis and the random coordinate
        This will create 2 boxes from the original box passed in as an argument.
        params:
            axes: The axis to split on
            c1: The random coordinate to split on
            box: The original box to split on
        """
        if axes == Axes.ROW:
            b0 = Box(
                box.row_start,
                c1,
                box.col_start,
                box.col_stop,
                box.depth_start,
                box.depth_stop,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            b1 = Box(
                c1,
                box.row_stop,
                box.col_start,
                box.col_stop,
                box.depth_start,
                box.depth_stop,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            return [b0, b1]
        if axes == Axes.COL:
            b0 = Box(
                box.row_start,
                box.row_stop,
                box.col_start,
                c1,
                box.depth_start,
                box.depth_stop,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            b1 = Box(
                box.row_start,
                box.row_stop,
                c1,
                box.col_stop,
                box.depth_start,
                box.depth_stop,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            return [b0, b1]
        if axes == Axes.DEPTH:
            b0 = Box(
                box.row_start,
                box.row_stop,
                box.col_start,
                box.col_stop,
                box.depth_start,
                c1,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            b1 = Box(
                box.row_start,
                box.row_stop,
                box.col_start,
                box.col_stop,
                c1,
                box.depth_stop,
                distribution=self.distribution,
                distribution_args=self.distribution_args,
                name=self.name,
            )
            return [b0, b1]

    def spawn_children(self, min_size, mode, map=None) -> List[Box] | Tuple | None:
        """spawn subboxes from <self>"""
        if self.area() < min_size:
            return ()

        # if we have tabular data, we are of the form (1, x) so we grab 3 random numbers from the
        # distribution and partition with those
        if mode == "tabular" or mode == "spectral":
            return self.__1d_parts()

        # we have a 2 dimension image
        if mode == "RGB" or mode == "L":
            return self.__2d_parts(map=map)

        # we have a 3 dimension data
        if mode == "voxel":
            return self.__3d_parts(map=map)

        return []

    def area(self) -> int:
        """returns the area of a box"""

        if self.depth_start is not None and self.depth_stop is not None:
            return (
                (self.row_stop - self.row_start)
                * (self.col_stop - self.col_start)
                * (self.depth_stop - self.depth_start)
            )
        else:
            if self.row_start == 0 and self.row_stop == 0:
                return self.col_stop - self.col_start
            return (self.row_stop - self.row_start) * (self.col_stop - self.col_start)


class Box(BoxInternal, NodeMixin):
    """An object that represents a box, a set of
    boxes forms an occlusion"""

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        row_start,
        row_stop,
        col_start,
        col_stop,
        depth_start=None,
        depth_stop=None,
        distribution=None,
        distribution_args=None,
        name="",
        parent=None,
        children=None,
    ) -> None:
        super().__init__(
            row_start,
            row_stop,
            col_start,
            col_stop,
            depth_start,
            depth_stop,
            distribution,
            distribution_args,
            name,
        )
        self.parent = parent
        # anytree nodemixin needs every field to be a tuple, and NoneType object is not iterable,
        # so to get around this limitaton, we set children to be an empty tuple if the box has not spawned any subboxes
        self.children = () if children is None else children

    def add_children_to_tree(self, min_size, mode, map):
        """adds a list of boxes (children) to box tree"""
        if self.children == ():
            try:
                self.children = self.spawn_children(min_size, mode, map)
            except TypeError:
                self.children = ()


def initialise_tree(
    r_lim,
    c_lim,
    distribution,
    distribution_args,
    r_start=0,
    c_start=0,
    d_start=None,
    d_lim=None,
) -> Box:
    """initialise box tree with root node, the whole image"""
    if d_lim is not None and d_start is None:
        d_start = 0
    return Box(
        r_start,
        r_lim,
        c_start,
        c_lim,
        depth_start=d_start,
        depth_stop=d_lim,
        distribution=distribution,
        distribution_args=distribution_args,
        name="R",
    )


def show_tree(tree):
    """Print the box tree to the terminal."""
    print(RenderTree(tree))


def average_box_size(tree, d) -> float:
    """Calculate the average box size at depth <d> of <tree>."""
    areas = [
        [node.area() for node in children] for children in LevelOrderGroupIter(tree)
    ]
    try:
        return np.mean(areas[d], axis=0)
    except IndexError:
        return 0.0


def box_dimensions(
    box: Box,
) -> tuple[int, int, int, int, int, int] | tuple[int, int, int, int]:
    """Returns box dimensions as a 4-tuple or 6-tuple depending on whether the box is 2D or 3D.

    @param box: Box
    @return (int, int, int, int) | (int, int, int, int, int, int)
    """
    if box.depth_start is not None and box.depth_stop is not None:
        return (
            box.row_start,
            box.row_stop,
            box.col_start,
            box.col_stop,
            box.depth_start,
            box.depth_stop,
        )
    else:
        return (box.row_start, box.row_stop, box.col_start, box.col_stop)


def boxes_name_and_dimensions(boxes: List[Box]):
    """Returns a list of all boxes and their dimensions"""
    return [(box.name, box_dimensions(box)) for box in boxes]
