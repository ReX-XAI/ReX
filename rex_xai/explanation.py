#!/usr/bin/env python3
# from __future__ import annotations
"""main logical entrypoint for ReX."""

import copy
import os
import sys
import time
from typing import Tuple, List, Union

from scipy.io import loadmat
import numpy as np
import torch as tt
from PIL import Image
from sqlalchemy.orm import Session
from tqdm import trange  # type: ignore

from rex_xai.config import CausalArgs
from rex_xai.database import update_database
from rex_xai.evaluation import Evaluation
from rex_xai.multi_explanation import MultiExplanation
from rex_xai.extraction import Explanation
from rex_xai.input_data import Data
from rex_xai.logger import logger
from rex_xai.onnx import get_prediction_function
from rex_xai.resp_maps import ResponsibilityMaps
from rex_xai.responsibility import causal_explanation
from rex_xai.prediction import Prediction
from rex_xai._utils import Strategy, ReXScriptError


def try_preprocess(args: CausalArgs, model_shape: Tuple[int], device: tt.device):
    """Makes an attempt to preprocess input data as required for the model.

    Data preprocessing is based on file extension and (possibly) user-defined mode.
    File extensions in ``[".jpg", ".jpeg", ".png", ".tif", ".tiff"]`` are treated
    as images, ".npy" and ".mat" are treated as Numpy arrays, and ".nii" are treated as nifti files.
    For any other file extension, we create a ``Data`` object without pre-processing.

    Args:
        args: configuration values for ReX
        model_shape: shape of the input tensor of the model, as returned by
            :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`

    Returns:
      Data: the processed input data
    """
    _, ext = os.path.splitext(args.path)
    # args.path is an image
    if ext.lower() in [".jpg", ".jpeg", ".png", ".tif", ".tiff"]:
        # a simple sanity check: spectral_occlusion is not suitable for images
        # so remind the user to update rex.toml and set mask_value to 0 in the meantime
        try:
            if args.mask_value in ("tabular", "spectral"):  # type: ignore
                logger.warning(
                    f"{args.mask_value} is not suitable for images. Changing mask_value to 0"
                )
                args.mask_value = 0
        except AttributeError:
            pass

        img = Image.open(args.path)
        if img.mode == "RGBA":
            logger.warning("RGBA input image provided, converting to RGB")
            img = img.convert("RGB")

        data = Data(img, model_shape, device)
        data.generic_image_preprocess(means=args.means, stds=args.stds, norm=args.norm)

    # a numpy "npy" array or matlab "mat" file
    elif ext in (".npy", ".mat"):
        if args.mode in ("tabular", "spectral"):
            if ext == ".mat":
                raw_data = np.load(loadmat(args.path)["val"])
            else:
                raw_data = np.load(args.path)
            data = Data(raw_data, model_shape, mode=args.mode, device=device)
            data.data = data.generic_tab_preprocess()
        else:
            logger.fatal("we do not generically handle this datatype")
            return NotImplemented
    # nifti files for 3d data
    elif ext in ".nii":
        logger.fatal("we do not (yet) handle nifti files generically")
        return NotImplemented
    else:
        # we don't know what to do!
        data = Data(args.path, model_shape, mode=args.mode, device=device)

    return data


def load_and_preprocess_data(
    model_shape: Tuple[int], device: tt.device, args: CausalArgs
):
    """Loads input data from filepath and does preprocessing.

    Uses a custom preprocesssing function if this is defined in ``args.script.preprocess``,
    otherwise :py:func:`~rex_xai.explanation.try_preprocess()`.

    Args:
        model_shape: shape of the input tensor of the model, as returned by
            :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        args: configuration values for ReX

    Returns:
        Data: the processed input data

    """
    if args.script is not None:
        if hasattr(args.script, "preprocess"):
            data = args.script.preprocess(
                args.path, model_shape, device, mode=args.mode
            )
        else:
            raise ReXScriptError(
                f"{args.script_location} is missing a preprocess() function"
            )
    else:
        # no custom preprocessing, so we make our best guess as to what to do
        data = try_preprocess(args, model_shape, device)

    return data


def predict_target(data: Data, prediction_func) -> Prediction:
    """Predicts classification of input data, using given prediction function.

    Uses ``prediction_func`` to identify the classification of the input data and return
    this as the target classification for ReX.

    Args:
        data: processed input data object
        args: configuration values for ReX
        prediction_func: prediction function for the model

    Returns:
        Prediction: the predicted target classification and confidence
    """
    target = prediction_func(data.data, None)

    if isinstance(target, list):
        target = target[0]

    if target is not None:
        logger.info(
            "image classified as %s with %f confidence",
            target.classification,
            target.confidence,
        )
    else:
        logger.warning("no target found")
        sys.exit(-1)

    return target


def calculate_responsibility(
    data: Data, args: CausalArgs, prediction_func, keep_all_maps=False
) -> tuple[ResponsibilityMaps, dict]:
    """Calculates ResponsibilityMaps for input data using given args.

    Runs :py:func:`~rex_xai.responsibility.causal_explanation` for ``args.iters`` iterations,
    and returns a ResponsibilityMaps object and a dictionary containing some statistics about
    the calculation process.
    The ResponsibilityMaps object by default only includes the responsibility map that matches
    the classification of the input data. Set ``keep_all_maps`` to ``True`` to retain all maps.

    Args:
        data: processed input data object
        args: configuration values for ReX
        prediction_func: prediction function for the model
        keep_all_maps: whether to retain all :py:class:`~rex_xai.resp_maps.ResponsibilityMaps`,
            or just the one that matches the target classification.

    Returns:
        tuple containing

        - ResponsibilityMaps: ResponsibilityMaps for the given data, prediction function, and args.
        - dict: statistics for the call of this function that generated the ResponsibilityMaps object
    """

    if data.target is None or data.target.classification is None:
        raise ValueError(
            "No target classification found. Please run `predict_target` before running `calculate_responsibility`."
        )

    maps = ResponsibilityMaps()
    if data.model_height is not None:
        maps.new_map(
            data.target.classification,
            data.model_height,
            data.model_width,
            data.model_depth,
        )
    else:
        maps.new_map(data.target.classification, data.model_height, data.model_width)

    total_passing: int = 0
    total_failing: int = 0
    max_depth_reached: int = 0
    avg_box_size: float = 0.0

    if args.iters > 0:
        for i in trange(args.iters, disable=not args.progress_bar):
            (
                local_maps,
                passing,
                failing,
                depth_reached,
                avg_bs,  # average box size
            ) = causal_explanation(
                i + 1,
                data,
                args,
                prediction_func,
                current_map=maps.get(data.target.classification),
            )

            total_passing += passing
            total_failing += failing
            max_depth_reached = max(max_depth_reached, depth_reached)
            avg_box_size += avg_bs
            # TODO this needs to be smarter. If we only have shallow penetration, then
            # this is doing us a disservice. Perhaps leave merging until after completion
            # of all iterations. Might potentially use a lot of memeory though
            if depth_reached > 1:
                maps.merge(local_maps)

    avg_box_size /= args.iters

    logger.info(
        "Total Passing: %d, Total Failing %d, Max Depth Reached %d, Avg Box Size %f",
        total_passing,
        total_failing,
        max_depth_reached,
        avg_box_size,
    )

    run_stats = {
        "total_passing": total_passing,
        "total_failing": total_failing,
        "max_depth_reached": max_depth_reached,
        "avg_box_size": avg_box_size,
    }

    return maps, run_stats


def analyze(exp: Explanation, data_mode: str | None):
    """Analyzes an Explanation.

    Analyzes the area ratio, entropy difference, insertion and deletion curves for an
    Explanation object, prints them, and returns them.

    Args:
        exp: Explanation object as returned by :py:func:`~rex_xai.explanation._explanation`
        data_mode: Mode of the input data. Entropy difference is only calculated if ``data_mode``
            is one of ["RGB", "L"].

    Returns:
        tuple containing

        - area (float)
        - entropy_diff (float)
        - insertion_curve (float)
        - deletion_curve (float)

    """
    eval = Evaluation(exp)
    rat = eval.ratio()
    ent = None
    max_ent = None
    if data_mode in ("RGB", "RGBA", "L"):
        be, ae = eval.entropy_loss()  # type: ignore
        ent = be - ae
    elif data_mode in ("spectral", "tabular"):
        ent, max_ent = eval.spectral_entropy()

    iauc, dauc = eval.insertion_deletion_curve(
        exp.prediction_func, normalise=exp.args.normalise_curves
    )

    analysis_results = {
        "area": rat,
        "entropy": ent,
        "max_entropy": max_ent,
        "insertion_curve": iauc,
        "deletion_curve": dauc,
    }

    return analysis_results


def _explanation(
    args: CausalArgs,
    model_shape: Tuple[int],
    prediction_func,
    device: tt.device,
    db: Session | None = None,
    path=None,
):
    """Takes a CausalArgs object and model information and returns a Explanation.

    Takes a CausalArgs object, model shape and prediction function and returns an Explanation.
    Depending on the input ``args``, optionally produces output plots, analyses the output
    explanation, and/or writes results to a database.

    Args:
        args: configuration values for ReX
        model_shape: shape of the input tensor of the model, as returned by :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        prediction_func: as returned by :py:func:`~rex_xai.explanation.get_prediction_func_from_args()`
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        db: None or as returned by :py:func:`~rex_xai.database.initialise_rex_db()`

    Returns:
        Explanation:
            An :py:class:`~rex_xai.extraction.Explanation` object containing the causal reponsibility explanation
            calculated using the given ``args``.

    """
    data = load_and_preprocess_data(model_shape, device, args)
    data.set_mask_value(args.mask_value)
    logger.debug(
        "args.mask_value is %s, data.mask_value is %s", args.mask_value, data.mask_value
    )

    data.target = predict_target(data, prediction_func)

    start = time.time()

    logger.info("Calculating responsibility map")
    resp_object, run_stats = calculate_responsibility(data, args, prediction_func)

    logger.info("Extracting explanation from responsibility map")
    clauses = None
    if args.strategy in (Strategy.MultiSpotlight, Strategy.Contrastive):
        exp = MultiExplanation(resp_object, prediction_func, data, args, run_stats)
        exp.extract()

        if args.strategy == Strategy.Contrastive and args.permitted_overlap != 1.0:
            logger.warning(
                "contrastive explanations require a permitted overlap of 1.0, so setting this now"
            )
            args.permitted_overlap = 1.0

        clauses = exp.separate_by(args.permitted_overlap)
        logger.info(f"found the following sets of explanations {clauses}")

        if args.strategy == Strategy.Contrastive:
            clauses = exp.contrastive(clauses)
            args.multi_style = "contrastive"
    else:
        exp = Explanation(resp_object, prediction_func, data, args, run_stats)
        exp.extract(args.strategy)

    if args.analyze:
        logger.info("Analysing explanation")
        results = analyze(exp, data.mode)
        if data.mode == "spectral":
            print(
                f"INFO:ReX:classification {exp.data.target.classification}, area {results['area']}, responsibility entropy {results['entropy']},",  # type: ignore
                f"max entropy {results['max_entropy']}",
                f"insertion curve {results['insertion_curve']}, deletion curve {results['deletion_curve']}",
            )
        else:
            print(
                f"INFO:ReX:classification {exp.data.target.classification}, area {results['area']}, entropy {results['entropy']},",  # type: ignore
                f"insertion curve {results['insertion_curve']}, deletion curve {results['deletion_curve']}",
            )

    end = time.time()
    time_taken = end - start
    logger.info(f"Time taken: {time_taken:.2f}s")

    if args.surface is not None:
        if path is not None:
            pass
        elif args.surface == "show":
            path = None
        else:
            path = args.surface
        exp.surface_plot(path)

    if args.heatmap is not None:
        if args.heatmap == "show":
            path = None
        else:
            path = args.heatmap
        exp.heatmap_plot(path)

    if args.output is not None:
        if path is None:
            if args.output == "show":
                path = None
            else:
                path = args.output
        exp.save(path, clauses=clauses)

    if db is not None:
        if args.strategy == Strategy.MultiSpotlight:
            logger.info("writing multiple explanations to database")
            update_database(
                db,
                exp,
                time_taken,
                multi=True,
            )
        else:
            logger.info("writing to database")
            update_database(
                db,
                exp,
                time_taken,
            )

    return exp


def get_prediction_func_from_args(args: CausalArgs):
    """Takes a CausalArgs object and gets the prediction function and model shape.

    If ``args.script`` specifies a prediction function and model shape, returns these.
    Otherwise gets the prediction function and model shape from the provided model
    file.

    Args:
        args: configuration values for ReX

    Returns:
        tuple containing

        - ``prediction_func``
        - ``model_shape``

    Raises:
        RuntimeError: if an onnx inference instance cannot be created from the provided model file.

    """
    if hasattr(args.script, "prediction_function") and hasattr(
        args.script, "model_shape"
    ):
        prediction_func = args.script.prediction_function  # type: ignore
        model_shape = args.script.model_shape()  # type: ignore
    else:
        ps = get_prediction_function(args)
        if ps is None:
            raise RuntimeError("Unable to create an onnx inference instance")
        else:
            prediction_func, model_shape = ps

    return prediction_func, model_shape


def explanation(
    args: CausalArgs, device: tt.device, db: Session | None = None
) -> Union[Explanation, List[Explanation]]:
    """Takes a CausalArgs object and returns a Explanation.

    Takes a CausalArgs object and returns either an Explanation, or a list of Explanations
    if the input ``args.path`` is a directory rather than a path to a single file.

    Args:
        args: configuration values for ReX
        device: as returned by :py:func:`~rex_xai._utils.get_device()`
        db: None or as returned by :py:func:`~rex_xai.database.initialise_rex_db()`

    Returns:
        Explanation:
            An :py:class:`~rex_xai.extraction.Explanation` object containing the causal reponsibility explanation
            calculated using the given ``args``.

    """

    prediction_func, model_shape = get_prediction_func_from_args(args)

    if isinstance(model_shape[0], int) and model_shape[0] < args.batch_size:
        logger.warning(
            f"Resetting batch size to size of model's first axis: {model_shape[0]}"
        )
        args.batch_size = model_shape[0]

    # directory of data to process
    if os.path.isdir(args.path):
        explanations = []
        dir = args.path
        path = None
        for dir, _, files in os.walk(args.path):
            for f in files:
                to_process = os.path.join(dir, f)
                logger.info("processing %s", to_process)
                current_args = copy.copy(args)
                current_args.path = to_process
                if args.output is not None and args.output != "show":
                    out_dir = os.path.dirname(args.output)
                    name, ext = os.path.splitext(args.output)
                    fname, _ = os.path.splitext(f)
                    path = f"{out_dir}_{fname}_{name}{ext}"
                exp = _explanation(
                    current_args,
                    model_shape,
                    prediction_func,
                    device,
                    db,
                    path=path,
                )
                explanations.append(exp)
        return explanations

    else:
        # a single explanation
        return _explanation(args, model_shape, prediction_func, device, db)
