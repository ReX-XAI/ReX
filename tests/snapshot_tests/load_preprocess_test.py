from rex_xai.explanation import (
    load_and_preprocess_data,
    try_preprocess,
)


def test_preprocess(args, model_shape, cpu_device, snapshot):
    data = load_and_preprocess_data(model_shape, cpu_device, args)
    assert data == snapshot


def test_preprocess_custom(args_custom, model_shape, cpu_device, snapshot):
    data = load_and_preprocess_data(model_shape, cpu_device, args_custom)
    assert data == snapshot


def test_preprocess_spectral_mask_on_image_returns_warning(
    args, model_shape, cpu_device, snapshot, caplog
):
    args.mask_value = "spectral"
    data = try_preprocess(args, model_shape, device=cpu_device)

    assert args.mask_value == 0
    assert (
        caplog.records[0].msg
        == "spectral is not suitable for images. Changing mask_value to 0"
    )
    assert data == snapshot


def test_preprocess_npy(args, model_shape, cpu_device, snapshot, caplog):
    args.path = "tests/test_data/DoublePeakClass 0 Mean.npy"
    try_preprocess(args, model_shape, device=cpu_device)

    assert caplog.records[0].msg == "we do not generically handle this datatype"

    args.mode = "tabular"
    data = try_preprocess(args, model_shape, device=cpu_device)
    assert data == snapshot
