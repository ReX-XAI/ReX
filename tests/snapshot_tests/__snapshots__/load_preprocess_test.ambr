# serializer version: 1
# name: test_preprocess
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
# name: test_preprocess_custom
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
# name: test_preprocess_npy
  Data: tabular, ['N', 3, 224, 224], None, None, 1, None
# ---
# name: test_preprocess_spectral_mask_on_image_returns_warning
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
